pkg/trace/agent/agent.go

```diff
@@ -237,50 +237,41 @@ func (a *Agent) Process(t pb.Trace) {
 		pt.Env = tenv
 	}
 
-	go func() {
-		defer watchdog.LogOnPanic()
-		a.ServiceExtractor.Process(pt.WeightedTrace)
-	}()
-
-	go func(pt ProcessedTrace) {
-		defer watchdog.LogOnPanic()
-		// Everything is sent to concentrator for stats, regardless of sampling.
-		a.Concentrator.Add(&stats.Input{
-			Trace:     pt.WeightedTrace,
-			Sublayers: pt.Sublayers,
-			Env:       pt.Env,
-		})
-	}(pt)
+	a.ServiceExtractor.Process(pt.WeightedTrace)
+
+	// Everything is sent to concentrator for stats, regardless of sampling.
+	a.Concentrator.Add(&stats.Input{
+		Trace:     pt.WeightedTrace,
+		Sublayers: pt.Sublayers,
+		Env:       pt.Env,
+	})
 
 	// Don't go through sampling for < 0 priority traces
 	if priority < 0 {
 		return
 	}
-	// Run both full trace sampling and transaction extraction in another goroutine.
-	go func(pt ProcessedTrace) {
-		defer watchdog.LogOnPanic()
 
-		tracePkg := writer.TracePackage{}
+	// Run both full trace sampling and transaction extraction in another goroutine.
+	tracePkg := writer.TracePackage{}
 
-		sampled, rate := a.sample(pt)
+	sampled, rate := a.sample(pt)
 
-		if sampled {
-			pt.Sampled = sampled
-			sampler.AddGlobalRate(pt.Root, rate)
-			tracePkg.Trace = pt.Trace
-		}
+	if sampled {
+		pt.Sampled = sampled
+		sampler.AddGlobalRate(pt.Root, rate)
+		tracePkg.Trace = pt.Trace
+	}
 
-		// NOTE: Events can be processed on non-sampled traces.
-		events, numExtracted := a.EventProcessor.Process(pt.Root, pt.Trace)
-		tracePkg.Events = events
+	// NOTE: Events can be processed on non-sampled traces.
+	events, numExtracted := a.EventProcessor.Process(pt.Root, pt.Trace)
+	tracePkg.Events = events
 
-		atomic.AddInt64(&ts.EventsExtracted, int64(numExtracted))
-		atomic.AddInt64(&ts.EventsSampled, int64(len(tracePkg.Events)))
+	atomic.AddInt64(&ts.EventsExtracted, int64(numExtracted))
+	atomic.AddInt64(&ts.EventsSampled, int64(len(tracePkg.Events)))
 
-		if !tracePkg.Empty() {
-			a.tracePkgChan <- &tracePkg
-		}
-	}(pt)
+	if !tracePkg.Empty() {
+		a.tracePkgChan <- &tracePkg
```

Comment:
Not to reviewers:

Because the goroutine has been removed, this is now a blocking call. This means that if the writer is slow, the [main loop](https://github.com/DataDog/datadog-agent/blob/b4d1e4cc2759dd771a745643d0205ec961afe846/pkg/trace/agent/agent.go#L136-L138) will be blocking and [traces will be dropped](https://github.com/DataDog/datadog-agent/blob/b4d1e4cc2759dd771a745643d0205ec961afe846/pkg/trace/api/api.go#L253-L261).

The other version (which we had before) would start a new goroutine for each processed trace, meaning that it could potentially start an indefinite amount of goroutines on a slow writer, likely causing memory leaks and OOM kills in some situations.

I tend to lean in the favor of this version as a potential release candidate.

Reply:
Should `tracePkgChan` become a buffered channel to smooth resource usage / blocking calls?

Reply:
It could, after my [new idea](https://github.com/DataDog/datadog-agent/pull/2995#issuecomment-461041487) gets implemented, it might not matter that much.